src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/KeyValuePairConverter.cs

```diff
@@ -38,107 +34,66 @@ internal override void Initialize(JsonSerializerOptions options)
                 _keyName = namingPolicy.ConvertName(KeyNameCLR);
                 _valueName = namingPolicy.ConvertName(ValueNameCLR);
 
-                if (_keyName == null || _valueName == null)
-                {
-                    ThrowHelper.ThrowInvalidOperationException_NamingPolicyReturnNull(namingPolicy);
-                }
+                // Validation for the naming policy will occur during JsonPropertyInfo creation.
             }
 
-            JavaScriptEncoder? encoder = options.Encoder;
-            _keyNameEncoded = JsonEncodedText.Encode(_keyName, encoder);
-            _valueNameEncoded = JsonEncodedText.Encode(_valueName, encoder);
+            ConstructorInfo = TypeToConvert.GetConstructor(new[] { s_keyType, s_valueType });
+            Debug.Assert(ConstructorInfo != null);
         }
 
-        internal override bool OnTryRead(
-            ref Utf8JsonReader reader,
-            Type typeToConvert, JsonSerializerOptions options,
+        /// <summary>
+        /// Lookup the constructor parameter given its name in the reader.
+        /// </summary>
+        protected override bool TryLookupConstructorParameter(
             ref ReadStack state,
-            out KeyValuePair<TKey, TValue> value)
+            ref Utf8JsonReader reader,
+            JsonSerializerOptions options,
+            out JsonParameterInfo? jsonParameterInfo)
         {
-            if (reader.TokenType != JsonTokenType.StartObject)
-            {
-                ThrowHelper.ThrowJsonException();
-            }
+            JsonClassInfo classInfo = state.Current.JsonClassInfo;
 
-            TKey k = default!;
-            bool keySet = false;
-
-            TValue v = default!;
-            bool valueSet = false;
-
-            // Get the first property.
-            reader.ReadWithVerify();
-            if (reader.TokenType != JsonTokenType.PropertyName)
-            {
-                ThrowHelper.ThrowJsonException();
-            }
+            Debug.Assert(state.Current.JsonClassInfo.ClassType == ClassType.Object);
+            Debug.Assert(_keyName != null);
+            Debug.Assert(_valueName != null);
 
             bool caseInsensitiveMatch = options.PropertyNameCaseInsensitive;
 
             string propertyName = reader.GetString()!;
-            if (FoundKeyProperty(propertyName, caseInsensitiveMatch))
-            {
-                reader.ReadWithVerify();
-                k = JsonSerializer.Deserialize<TKey>(ref reader, options, ref state, _keyName);
-                keySet = true;
-            }
-            else if (FoundValueProperty(propertyName, caseInsensitiveMatch))
-            {
-                reader.ReadWithVerify();
-                v = JsonSerializer.Deserialize<TValue>(ref reader, options, ref state, _valueName);
-                valueSet = true;
-            }
-            else
+            if (!state.Current.CtorArgumentState!.FoundKey &&
```

Comment:
When can `CtorArgumentState` be null? If its only when the `*WithParameterizedConstructorConverter` is not used, then maybe you could `Debug.Assert(CtorArgumentState != null)` at the beginning of the method in order to remove all the `!`.